### PR TITLE
Resolve or ignore Ruff and Pylint warnings

### PR DESCRIPTION
**Pull request by mattt**
Ignoring warnings related to `TODO` and `FIXME` comments:

```
python -m pylint python/ --disable W0511
```

This is a lot, but I think it would help us to ratchet up code quality ahead of merging async.

### CODE REVIEW

File: python/cog/predictor.py [hunk 1]
```diff
@@ -391,8 +394,8 @@ def get_output_type(predictor: BasePredictor) -> Type[BaseModel]:
 
     input_types = get_type_hints(predict)
 
-    OutputType = input_types.pop("return", None)
-    if OutputType is None:
+    OutputType = input_types.pop("return", None)  # pylint: disable=invalid-name
+    if not OutputType:
```
Comment (mattt) [inline python/cog/predictor.py:398]: Duplicate of #1824. Will resolve that there first and rebase.

File: python/cog/predictor.py [hunk 2]
```diff
@@ -492,7 +495,7 @@ def get_training_output_type(predictor: BasePredictor) -> Type[BaseModel]:
     train = get_train(predictor)
 
     input_types = get_type_hints(train)
-    TrainingOutputType = input_types.pop("return", None)
+    TrainingOutputType = input_types.pop("return", None)  # pylint: disable=invalid-name
```
Comment (mattt) [inline python/cog/predictor.py:498]: We should apply the same pattern as suggested in https://github.com/replicate/cog/pull/1824#discussion_r1691339203 here as well.

File: python/cog/server/probes.py [hunk 3]
```diff
@@ -24,9 +24,10 @@ def __init__(self, root: PathLike = None) -> None:
             self._root.mkdir(exist_ok=True, parents=True)
         except OSError:
             log.error(
-                f"Failed to create cog runtime state directory ({self._root}). "
+                "Failed to create cog runtime state directory (%s). "
                 "Does it already exist and is a file? Does the user running cog "
-                "have permissions?"
+                "have permissions?",
+                self._root,
```
Comment (mattt) [inline python/cog/server/probes.py:30]: Flagging this as a nontrivial change.

File: python/cog/server/runner.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 import structlog
 from attrs import define
 from requests.adapters import HTTPAdapter
-from requests.packages.urllib3.util.retry import Retry  # type: ignore
+from urllib3.util.retry import Retry
```
Comment (mattt) [inline python/cog/server/runner.py:14]: TIL. This is wild.

https://github.com/psf/requests/blob/0e322af87745eff34caffe4df68456ebc20d9068/src/requests/packages.py#L6

File: python/cog/types.py [hunk 5]
```diff
@@ -150,9 +149,11 @@ def __init__(self, *, source: str, filename: str, fileobj: io.IOBase) -> None:
...
     def convert(self) -> Path:
         if self._path is None:
-            dest = tempfile.NamedTemporaryFile(suffix=self.filename, delete=False)
-            shutil.copyfileobj(self.fileobj, dest)
-            self._path = Path(dest.name)
+            with tempfile.NamedTemporaryFile(
+                suffix=self.filename, delete=False
+            ) as dest:
+                shutil.copyfileobj(self.fileobj, dest)
+                self._path = Path(dest.name)
```
Comment (mattt) [inline python/cog/types.py:156]: Double-check that use of context manager `NamedTemporaryFile` with `delete=False` does what we expect.

File: python/cog/types.py [hunk 6]
```diff
@@ -263,6 +261,9 @@ def get_filename(url: str) -> str:
 
 
 class ConcatenateIterator(Iterator[Item]):
+    def __next__(self) -> Item:
+        raise NotImplementedError("This is a stub class for typing purposes only")
```
Comment (mattt) [inline python/cog/types.py:265]: Is this right?

**Comment by technillogue** [comment]
are we using both pylint and ruff now?

File: python/cog/types.py (hunk 5 above)
Comment (mattt) [reply python/cog/types.py:156]: On review, I don't think this is equivalent. Backing out of this change.

**Comment by mattt** [comment]
> are we using both pylint and ruff now?

@technillogue Yes. And Pyright, too. Taking a belt and suspenders and chainmail approach, I suppose.

File: python/cog/types.py (hunk 6 above)
Comment (mattt) [reply python/cog/types.py:263]: This isn't obviously correct, and not worth the heartache. Backing out and ignoring instead.

**Pull request merged**